### PR TITLE
[SPARK-27774][CORE][MLLIB] Avoid hardcoded configs

### PR DESCRIPTION
**Pull request by wenxuanguan**
## What changes were proposed in this pull request?

avoid hardcoded configs in `SparkConf` and `SparkSubmit` and test

## How was this patch tested?

N/A

### CODE REVIEW

File: docs/structured-streaming-kafka-integration.md [hunk 1]
```diff
@@ -441,7 +441,7 @@ Apache Kafka only supports at least once write semantics. Consequently, when wri
 or Batch Queries---to Kafka, some records may be duplicated; this can happen, for example, if Kafka needs
 to retry a message that was not acknowledged by a Broker, even though that Broker received and wrote the message record.
 Structured Streaming cannot prevent such duplicates from occurring due to these Kafka write semantics. However,
-if writing the query is successful, then you can assume that the query output was written at least once. A possible
+if writing the query is successful, then you can assume that the query output was written exactly once. A possible
```
Comment (dongjoon-hyun) [inline docs/structured-streaming-kafka-integration.md]: Hi, @wenxuanguan . This looks wrong in this context. Could you explain your thought?

Comment (wenxuanguan) [reply docs/structured-streaming-kafka-integration.md]: Thanks for your reply.
I thought this describes the situation that query writes to kafka successfully and no records are duplicated. How about change to `So if writing the query is successful, then you can assume that the query output was written at least once`, which will not be confused by `However`

Comment (dongjoon-hyun) [reply docs/structured-streaming-kafka-integration.md]: The existing one looks okay to me. Let's remove this doc change from this PR.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Please revert the doc change and remove `[DOC]` and `fix kafka sink write semantics in document ` from the title.

BTW, @wenxuanguan . We usually don't put different stuffs into a single PR. Single PR had better have a single issue. Please make separate PRs for the orthogonal issues next time. Thanks.

File: docs/structured-streaming-kafka-integration.md (hunk 1 above)
Comment (wenxuanguan) [reply docs/structured-streaming-kafka-integration.md]: @dongjoon-hyun Thanks, done

**Comment by wenxuanguan** [comment]
@dongjoon-hyun Thanks for your reply.
I have changed title and description, and reverted the document change.
Thanks for your suggestion, and it really helps me.

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by srowen** [comment]
This is probably OK but are there more instances like this?

**Review by HyukjinKwon — APPROVED** [review]
Yes, looks fine but it would be better if we fix them more

**Comment by wenxuanguan** [comment]
> This is probably OK but are there more instances like this?

I found nothing so far except `SparkSubmit`, and I have fixed it.

File: core/src/main/scala/org/apache/spark/SparkConf.scala [hunk 2]
```diff
@@ -595,7 +595,7 @@ class SparkConf(loadDefaults: Boolean) extends Cloneable with Logging with Seria
     // it will almost always cause ExecutorLostFailure. See SPARK-22754.
     require(executorTimeoutThresholdMs > executorHeartbeatIntervalMs, "The value of " +
       s"${networkTimeout}=${executorTimeoutThresholdMs}ms must be no less than the value of " +
-      s"spark.executor.heartbeatInterval=${executorHeartbeatIntervalMs}ms.")
+      s"${EXECUTOR_HEARTBEAT_INTERVAL.key}=${executorHeartbeatIntervalMs}ms.")
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/SparkConf.scala]: Could you update `SparkContextSuite.scala`, too?
```
-      .set("spark.executor.heartbeatInterval", "1s")
+      .set(EXECUTOR_HEARTBEAT_INTERVAL.key, "1s")
```

File: core/src/main/scala/org/apache/spark/SparkConf.scala [hunk 3]
```diff
@@ -667,12 +667,12 @@ private[spark] object SparkConf extends Logging {
         translation = s => s"${s.toLong * 10}s")),
     REDUCER_MAX_SIZE_IN_FLIGHT.key -> Seq(
       AlternateConfig("spark.reducer.maxMbInFlight", "1.4")),
-    "spark.kryoserializer.buffer" -> Seq(
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/SparkConf.scala]: For these `spark.kryoserializer.buffer` and `spark.kryoserializer.buffer.max`, it seems there are too many instances.
```
core/src/test/scala/org/apache/spark/SparkConfSuite.scala:    assert(conf.getSizeAsKb("spark.kryoserializer.buffer") === 1100)
core/src/test/scala/org/apache/spark/storage/BlockManagerReplicationSuite.scala:  conf.set("spark.kryoserializer.buffer", "1m")
core/src/test/scala/org/apache/spark/storage/BlockManagerReplicationSuite.scala:  conf.set("spark.kryoserializer.buffer", "1m")
core/src/test/scala/org/apache/spark/storage/BlockManagerReplicationSuite.scala:  conf.set("spark.kryoserializer.buffer", "1m")
core/src/test/scala/org/apache/spark/storage/BlockManagerSuite.scala:  val serializer = new KryoSerializer(new SparkConf(false).set("spark.kryoserializer.buffer", "1m"))
core/src/test/scala/org/apache/spark/storage/BlockManagerSuite.scala:      .set("spark.kryoserializer.buffer", "1m")
core/src/test/scala/org/apache/spark/storage/MemoryStoreSuite.scala:  val serializer = new KryoSerializer(new SparkConf(false).set("spark.kryoserializer.buffer", "1m"))
examples/src/main/scala/org/apache/spark/examples/mllib/MovieLensALS.scala:        .set("spark.kryoserializer.buffer", "8m")
```

```
mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:        sc.conf.get("spark.kryoserializer.buffer.max", "64m"))
mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala:        spark.conf.get("spark.kryoserializer.buffer.max", "64m"))
```

Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/SparkConf.scala]: You can use `git grep` to search these instances, @wenxuanguan .

**Comment by wenxuanguan** [comment]
retest this please

**Comment by dongjoon-hyun** [comment]
@wenxuanguan . Thank you for improving the PR. Now, this PR seems need a JIRA issue because it touches many files. Could you file a Apache Spark JIRA issue? Please enumerate the configs in the JIRA which you replaced. Also, please use that JIRA id instead of `[MINOR]`.

File: core/src/main/scala/org/apache/spark/SparkConf.scala (hunk 3 above)
Comment (wenxuanguan) [reply core/src/main/scala/org/apache/spark/SparkConf.scala]: @dongjoon-hyun Sorry for delay. 
I have fix the test, and changed some configs with the same problem.

**Comment by wenxuanguan** [comment]
@dongjoon-hyun Jira SPARK-27774 has been made, and linked with this pr.

File: examples/src/main/scala/org/apache/spark/examples/mllib/MovieLensALS.scala [hunk 4]
```diff
@@ -24,6 +24,7 @@ import org.apache.log4j.{Level, Logger}
 import scopt.OptionParser
 
 import org.apache.spark.{SparkConf, SparkContext}
+import org.apache.spark.internal.config.Kryo._
```
Comment (srowen) [inline examples/src/main/scala/org/apache/spark/examples/mllib/MovieLensALS.scala]: Hm, I think we might not want to modify the examples actually. They demonstrate what users might run. But they won't be able to import internal classes like this, no?

Comment (wenxuanguan) [reply examples/src/main/scala/org/apache/spark/examples/mllib/MovieLensALS.scala]: @srowen That make sense, revert it.

**Comment by wenxuanguan** [comment]
retest this please

**Comment by wenxuanguan** [comment]
retest this please

**Comment by wenxuanguan** [comment]
retest this please

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by wenxuanguan** [comment]
@HyukjinKwon @srowen @dongjoon-hyun  Thank you for review

**Pull request closed without merging**